### PR TITLE
feat(docs-page): content-api calls

### PR DESCRIPTION
**Pull request by thiskevinwang**
🎟️ [Asana Task]()
🔍 [Preview Link](https://react-components-git-{branch-slug}-hashicorp.vercel.app)

---

## Description

For https://github.com/hashicorp/waypoint/pull/2344

This updates `docs-page` to request data from the [content api](https://github.com/hashicorp/mktg-content-workflows/)

Tests + TS Refactor:
- https://github.com/hashicorp/react-components/pull/360

### PR Checklist 🚀

Items in this checklist may not may not apply to your PR, but please consider each item carefully.

- [ ] Add Asana and Preview links above.
- [x] Conduct thorough self-review.
- [x] Add or update tests as appropriate.
- [ ] Conduct reasonable cross browser testing for both compatibility and responsive behavior (We have a [Sauce Labs](https://app.saucelabs.com/) account for this, if you don't have access, just ask!).
- [ ] Conduct reasonable accessibility review (use the [WAS](https://accessible.org/Web-Accessibility-Standards-WAS-2.pdf) as a guide or an [axe browser plugin](https://www.deque.com/axe/) until we establish more formal checks).
- [ ] Identify (in the description above) and document (add Asana tasks on [this board](https://app.asana.com/0/1100423001970639/list)) any technical debt that you're aware of, but are not addressing as part of this PR.

### CODE REVIEW

File: packages/docs-page/server.js [hunk 1]
```diff
@@ -3,18 +3,22 @@ import path from 'path'
...
-  loadVersionedDocument,
-  loadVersionedNavData,
-  getVersionFromPath,
-} from '@hashicorp/versioned-docs/server'
 import moize from 'moize'
-import { normalizeVersion } from '@hashicorp/versioned-docs/client'
+
 import renderPageMdx from './render-page-mdx'
+import { stripVersionFromPathParams, normalizeVersion } from './util'
+import { fetchNavData, fetchDocument, fetchVersionMetadataList } from './rpc'
```
Comment (BRKalow) [inline packages/docs-page/server.js:10]: nit: RPC doesn't seem like the correct term to use here...what do you think about `content-api` or just `api` for this folder?

File: packages/docs-page/server.js [hunk 2]
```diff
@@ -95,49 +97,78 @@ async function generateStaticProps({
...
+  if (ENABLE_VERSIONED_DOCS) {
+    // given: v0.5.x (latest), v0.4.x, v0.3.x
+    const [versionFromPath, paramsNoVersion] = stripVersionFromPathParams(
+      params[paramId]
+    )
+    // versionFromPath should realistically only ever be "latest" | "v0.4.x" | "v0.3.x"
+    // It could be v0.5.x if a use navigates directly to it.
 
-    versions = await loadVersionList(product.slug)
+    const versionMetadataList = await fetchVersionMetadataList(productSlug)
```
Comment (BRKalow) [inline packages/docs-page/server.js:111]: Let's use the cached version as this API response shouldn't change at all during a build!

```suggestion
    const versionMetadataList = await cachedFetchVersionMetadataList(productSlug)
```

File: packages/docs-page/server.js [hunk 3]
```diff
@@ -95,49 +97,78 @@ async function generateStaticProps({
...
       // Construct the githubFileUrl, used for "Edit this page" link
-      const githubFileUrl = `https://github.com/hashicorp/${product.slug}/blob/${doc.gitRef}/website/content/${doc.filePath}`
+      // https://github.com/hashicorp/waypoint/blob/main/website/content/commands/index.mdx
+      // https://github.com/hashicorp/waypoint/blob/e591452/website/content/commands/index.mdx
+      // https://github.com/hashicorp/waypoint/blob/e591452e6f9a155a038d45158e05d904ba006d37/website/content/commands/index.mdx
+      // const githubFileUrl = `https://github.com/hashicorp/${productSlug}/blob/${document.sha}/website/content/${doc.filePath}`
+
+      // UNSUPPORTED?
+      // Must be serializeable
+      const githubFileUrl = null
```
Comment (BRKalow) [inline packages/docs-page/server.js:166]: This should be achievable once we store the full filepath! ref: https://app.asana.com/0/1100423001970639/1201024261811837/f

File: packages/docs-page/server.js (hunk 1 above)
Comment (thiskevinwang) [reply packages/docs-page/server.js:10]: `content-api` it is!

File: packages/docs-page/server.js (hunk 3 above)
Comment (thiskevinwang) [reply packages/docs-page/server.js]: Left a Todo

File: packages/docs-page/server.js [hunk 4]
```diff
@@ -29,24 +37,23 @@ async function generateStaticPaths({
   paramId = DEFAULT_PARAM_ID,
   product,
   basePath,
-  currentVersion,
+  // currentVersion = 'latest',
```
Comment (BRKalow) [inline packages/docs-page/server.js:40]: Hmm, it seems like we can remove this!
```suggestion
```

**Review by BRKalow — COMMENTED** [review]
Looking good, nice tests! A few minor comments, I'm gonna do some quick validation before approval.

File: packages/docs-page/server.test.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  it.skip('should generate static props', async () => {
+    const input = {
+      navDataFile: NAV_DATA_FILE,
+      localContentDir: CONTENT_DIR,
+      product: { name: 'Waypoint', slug: 'waypoint' },
+      params: {},
+      basePath,
+    }
+    const res = await generateStaticProps(input as any)
+    expect(res).toMatchInlineSnapshot()
```
[the remark this answers is not in the document]
Comment (thiskevinwang) [reply packages/docs-page/server.test.ts:34]: `server.test.ts` tests are `.skip`'d for now. I was thinking to revisit asap/post-demo due it's complex logic

File: packages/docs-page/util/index.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
+const REGEX = /^v([0-9]+)\.([0-9]+)\.(x|[0-9]+)(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?(?:\+[0-9A-Za-z-]+)?$/i
+
+/**
+ * Given an array of strings, returns a tuple of
+ * - [0]: a version or 'latest'
+ * - [1]: an array without the version
+ * @param {string[]} pathParams
+ * @returns {[string, string[]]} [version, strippedParams]
+ */
+export const stripVersionFromPathParams = (pathParams = []) => {
```
[the remark this answers is not in the document]
Comment (thiskevinwang) [reply packages/docs-page/util/index.ts:10]: I'm not sure this assumption is future proof w/ dev-portal.

What if the catch-all route `[[...page]]` maps to -> `pathParams = ["terraform", "v1.0.0", "docs"]`

File: packages/docs-page/server.test.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
+import { generateStaticPaths, generateStaticProps } from './server'
+
+const NAV_DATA_FILE = 'data/commands-nav-data.json'
+const CONTENT_DIR = 'content/commands'
+const basePath = 'commands'
+
+describe('generateStaticPaths', () => {
```
[the remark this answers is not in the document]
Comment (thiskevinwang) [reply packages/docs-page/server.test.ts:7]: I have yet to fill in tests in `server.test.ts`!

File: packages/docs-page/util/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (BRKalow) [reply packages/docs-page/util/index.ts:10]: Fair enough!

File: packages/docs-page/server.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (thiskevinwang) [reply packages/docs-page/server.test.ts:34]: Incoming tests! 
- https://github.com/hashicorp/react-components/pull/360

File: packages/docs-page/server.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (thiskevinwang) [reply packages/docs-page/server.test.ts:7]: Incoming tests! https://github.com/hashicorp/react-components/pull/360

File: .changeset/four-buttons-knock.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+---
+'@hashicorp/react-docs-page': patch
```
Comment (BRKalow) [inline .changeset/four-buttons-knock.md:2]: Since we're changing functionality slightly:

```suggestion
'@hashicorp/react-docs-page': minor
```

**Review by BRKalow — APPROVED** [review]
:shipit:

**Pull request merged**